### PR TITLE
[AEA-903] multiplexer cancel recv send loops before disconnecting connections

### PR DESCRIPTION
**Pull request by solarw**
## Proposed changes

multiplexer cancel recv send loops before disconnecting connections

## Fixes

https://github.com/fetchai/agents-aea/issues/1738

fixes issue when multiplexer tries to handle messages in queues but connections alraedy disconnected.
now it stops message processing loops and disconnect connections

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [x] I have checked that the documentation about the `aea cli` tool works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: aea/multiplexer.py [hunk 1]
```diff
@@ -263,17 +263,36 @@ async def disconnect(self) -> None:
...
             try:
                 self.connection_status.set(ConnectionStates.disconnecting)
-                await asyncio.wait_for(self._disconnect_all(), timeout=60)
                 await asyncio.wait_for(self._stop(), timeout=60)
                 self.logger.debug("Multiplexer disconnected.")
             except (CancelledError, Exception):
                 self.logger.exception("Exception on disconnect:")
                 raise AEAConnectionError("Failed to disconnect the multiplexer.")
 
+    async def _stop_recv_send_loops(self) -> None:
```
Comment (DavidMinarsch) [inline aea/multiplexer.py:275]: Please don't abbreviate `recv` ..it is inconsistent with the code base (spelling things out!) and confusing. Relabel `receive`.

**Review by DavidMinarsch — APPROVED** [review]
LGTM, except some naming issues

**Comment by DavidMinarsch** [comment]
Please also add a test that ensures the cancel send receive happens before disconnect so we don't reintroduce the problem when refactoring

File: aea/multiplexer.py (hunk 1 above)
Comment (solarw) [reply aea/multiplexer.py]: done

**Comment by solarw** [comment]
> Please also add a test that ensures the cancel send receive happens before disconnect so we don't reintroduce the problem when refactoring

done

**Review by DavidMinarsch — APPROVED** [review]
LGTM

File: aea/multiplexer.py [hunk 2]
```diff
@@ -263,41 +263,58 @@ async def disconnect(self) -> None:
         async with self._lock:
             if self.connection_status.is_disconnected:
                 self.logger.debug("Multiplexer already disconnected.")
-                await asyncio.wait_for(self._stop(), timeout=60)
```
Comment (lrahmani) [inline aea/multiplexer.py:266]: @solarw why was this needed in the first place?
Just making the observation in case this was here for a reason (to handle an edge case for e.g.)

I understand that in the previous version disconnecting all connections and then cancelling background coroutines  *in this order* was problematic. However they were still both executed after a call of `disconnect()`.

**Review by lrahmani — COMMENTED** [review]
LGTM, just one question

File: aea/multiplexer.py (hunk 2 above)
Comment (solarw) [reply aea/multiplexer.py:266]: @lrahmani i think there is no reason to call stop second time if already disconnected, this produced me an error in tests after the refactoring. now all the tests are passing ok, so not sure it's required anymore.
